src/sentry/auth/providers/saml2.py

```diff
@@ -1,65 +1,134 @@
-            'nameid': auth.get_nameid(),
-            'session_index': auth.get_session_index()
-        }
+        attributes = auth.get_attributes()
+        nameid = auth.get_nameid()
 
-        helper.bind_state('data', data)
-        return helper.next_step()
-        #return HttpResponse("ACS endpoint")
+        email = self.retrieve_email(attributes, nameid, provider.config)
 
+        # Filter users based on the emails provided in the commits
+        user_emails = UserEmail.objects.filter(
+            email__iexact=email,
+            is_verified=True
+        ).order_by('id')
 
-class SAML2MetadataView(AuthView):
-    def dispatch(self, request, organization_slug):
-        try:
-            organization = Organization.objects.get(
-                slug=organization_slug
+        if user_emails:
+            users = User.objects.filter(
+                id__in=set((ue.user_id for ue in user_emails)),
+                is_active=True,
+                sentry_orgmember_set__organization_id=organization.id
             )
-        except Organization.DoesNotExist:
-            return self.redirect(reverse('sentry-login'))
+            if users:
+                if users.count() == 1:
+                    user = users[0]
+                    user.backend = settings.AUTHENTICATION_BACKENDS[0]
+                    if login(request,
+                             user,
+                             after_2fa=request.build_absolute_uri(),
+                             organization_id=organization.id):
+                        request.session['saml'] = {
+                            'nameid': nameid,
+                            'nameid_format': auth.get_nameid_format(),
+                            'session_index': auth.get_session_index()
+                        }
+                    return HttpResponseRedirect(get_login_redirect(request))
+                else:
+                    return HttpResponseServerError("Found several accounts related with %s on this organization" % email)
+            else:
+                return HttpResponseServerError("The user %s is not related with this organization" % email)
+        else:
+            return HttpResponseServerError("An user with a verified mail: %s does not exist" % email)
+
+    def retrieve_email(self, attributes, nameid, config):
+        possible_mail = None
+        if nameid and '@' in nameid:
+            possible_mail = nameid
+
+        if attributes and 'attribute_mapping' in config and 'attribute_mapping_email' in config['attribute_mapping']:
+            email_mapping = config['attribute_mapping']['attribute_mapping_email']
+            if email_mapping and email_mapping in attributes:
+                return attributes[email_mapping][0]
+            elif possible_mail:
+                return possible_mail
+            else:
+                raise Exception("Email was not provided by the IdP and is required in order to execute the SAML process")
```

Comment:
At the provider, I raised several exceptions when the plugin is not able to execute SSO...

I saw on the [oauth provider](https://github.com/getsentry/sentry/blob/master/src/sentry/auth/providers/oauth2.py#L213) that you also raised Exceptions... 

Maybe make sense to create something to catch provider's exceptions? or may we catch them on the provider and implement an error view to show them?